node/src/components/consensus/highway_core/highway/vertex.rs

```diff
@@ -28,6 +28,11 @@ where
     C: Context,
 {
     Unit(C::Hash),
+    /// The unit with its full panorama.
+    UnitWithPanorama(C::Hash),
+    /// A unit specified by sequence number and validator index, as seen by another unit, given by
+    /// hash.
+    UnitBySeqNum(u64, ValidatorIndex, C::Hash),
```

Comment:
We already have the skip list to find them easily — but only for honest validators, yes.

The problem with omitting the hash is the following situation:
Imagine `v < u` and I already have multiple forks by `v`'s creator in my protocol state, but the unit `u` doesn't happen to know about them: It cites `v` directly.
Now I send you `u`, and since you don't have it yet you request `v` from me — by sequence number!
Since your request doesn't contain `u`'s hash anymore I don't know what it refers to, and I have multiple forks by `v`'s creator, so I can't find the unit. We'd be stuck.

You could argue that I should just pick _any_ unit with the right sequence number. Maybe that would be better; but it would mean that I pick a random fork for every equivocator.

To avoid complicating this feature even further I'd prefer to leave it as is for now, and consider adding such an additional index separately in the future?

Reply:
> You could argue that I should just pick any unit with the right sequence number. Maybe that would be better; but it would mean that I pick a random fork for every equivocator.

Exactly! And if the end, the panorama's control hash doesn't match I will fall back to requesting the dependencies by hash.

Also, if you know about `V`'s forks but you don't mark `V` as faulty in your latest unit `u`, isn't that an invalid behavior and you will also end up marked as `Faulty`?

Reply:
`u` isn't necessarily my latest unit. It's not necessarily _my_ unit at all. Maybe I'm not even a validator!

I'm a peer who sent you (maybe even as a dependency of something else) `u` (which might be old, if you've just joined).
If `u` doesn't mark `v`'s creator as faulty, that's totally fine in general.
But I should be able to answer all requests for dependencies of `u` somehow.
If I give you the wrong unit with the right sequence number, that's also fine, because as you say, you'll request it by hash if it doesn't match in the end. It might just add more rounds of communication if I send you the wrong fork first and thereby force you to request the full hash-based panorama and the other fork by hash.

Anyway, maybe you're right; at least it would simplify the different kinds of dependencies. I can give it a try and make the change. Do you think it should be part of this PR?

Reply:
Thanks for the more detailed explanation, I missed some nuances originally.

> It might just add more rounds of communication if I send you the wrong fork first and thereby force you to request the full hash-based panorama and the other fork by hash.

The new approach, by design, adds more rounds of communication but that's only for the case where we have faulty/non-honest validators, and we're fine with that. This particular use-case we're discussing also involves equivocations.

I would prefer if we added it to this PR, so that it marks that piece of work as complete.